### PR TITLE
Stricter types

### PR DESCRIPTION
**Pull request by bershanskiy**
#### Summary

Refactor the type system a bit and remove extra type casts (thus making code more type-sound).

#### Test results and supporting details

This is pure refactoring.

#### Related issues

#16668

### CODE REVIEW

File: scripts/fix/browser-order.ts [hunk 1]
```diff
@@ -23,15 +23,17 @@ export const orderSupportBlock = (
   value: CompatStatement,
 ): CompatStatement => {
   if (key === '__compat') {
-    const support = Object.keys(value.support)
+    const support: CompatStatement['support'] = (
```
Comment (queengooborg) [inline scripts/fix/browser-order.ts:26]: Wouldn't it be better to do this instead?

```suggestion
    const support: SupportStatement = (
```

(Please feel free to ignore if this is just another one of those dumb TypeScript nits!)

**Comment by bershanskiy** [comment]
I dropped changes to the following files:
 - schemas/compat-data.schema.json
 - test/linter/test-consistency.ts
 - types/index.d.ts

They were changing interfaces too much which is probably bad now considering the BCD export `any` cast and that we might want to replace `"mirror"` statement with lack or anything at all.

**Review by queengooborg — APPROVED** [review]

**Comment by rebloor** [comment]
I seem to be having issues running test, which I'm told relate to this change:
```
npm test

> @mdn/browser-compat-data@5.1.1 test
> npm run format && npm run lint && npm run unittest


> @mdn/browser-compat-data@5.1.1 format
> npx eslint "**/*.ts" && npx prettier --check "**/*.md" "**/*.json"


/Users/…/Documents/repositories/…/browser-compat-data/scripts/release/build.test.ts
   6:8  warning  'bcd' is defined but never used               @typescript-eslint/no-unused-vars
   8:3  warning  'applyMirroring' is defined but never used    @typescript-eslint/no-unused-vars
   9:3  warning  'createDataBundle' is defined but never used  @typescript-eslint/no-unused-vars
  10:3  warning  'generateMeta' is defined but never used      @typescript-eslint/no-unused-vars

/Users/…/Documents/repositories/…/browser-compat-data/test/linter/test-obsolete.test.ts
  20:5  warning  '_' is defined but never used  @typescript-eslint/no-unused-vars

✖ 5 problems (0 errors, 5 warnings)

Checking formatting...
All matched files use Prettier code style!

> @mdn/browser-compat-data@5.1.1 lint
> ts-node test/lint

/Users/…/Documents/repositories/…/browser-compat-data/node_modules/ts-node/src/index.ts:843
    return new TSError(diagnosticText, diagnosticCodes, diagnostics);
           ^
TSError: ⨯ Unable to compile TypeScript:
test/linter/test-consistency.ts:486:13 - error TS2322: Type 'SupportStatement | undefined' is not assignable to type 'InternalSupportStatement'.
  Type 'undefined' is not assignable to type 'InternalSupportStatement'.

486         let browserData: InternalSupportStatement = compatData.support[browser];
                ~~~~~~~~~~~

    at createTSError (/Users/…/Documents/repositories/…/browser-compat-data/node_modules/ts-node/src/index.ts:843:12)
    at reportTSError (/Users/…/Documents/repositories/…/browser-compat-data/node_modules/ts-node/src/index.ts:847:19)
    at getOutput (/Users/…/Documents/repositories/…/browser-compat-data/node_modules/ts-node/src/index.ts:1057:36)
    at Object.compile (/Users/…/Documents/repositories/…/browser-compat-data/node_modules/ts-node/src/index.ts:1411:41)
    at transformSource (/Users/…/Documents/repositories/…/browser-compat-data/node_modules/ts-node/src/esm.ts:400:37)
    at /Users/…/Documents/repositories/…/browser-compat-data/node_modules/ts-node/src/esm.ts:278:53
    at async addShortCircuitFlag (/Users/…/Documents/repositories/…/browser-compat-data/node_modules/ts-node/src/esm.ts:409:15)
    at async ESMLoader.load (node:internal/modules/esm/loader:407:20)
    at async ESMLoader.moduleProvider (node:internal/modules/esm/loader:326:11) {
  diagnosticCodes: [ 2322 ]
}
```

**Comment by bershanskiy** [comment]
@rebloor Which commit are you on? I'm on current `main` `2776fb5` and can not reproduce.

**Comment by rebloor** [comment]
I pulled from Main a few minutes ago. Let me go and try main itself

**Comment by rebloor** [comment]
Sorry delete my earlier comment, was recommended to create a new issue

**Comment by rebloor** [comment]
I get the same issue off of Main, creating a new issue

**Comment by bershanskiy** [comment]
For anyone reading this: that issue is #16716, let's continue the discussion there.

**Pull request merged**